src/cmd/services/m3comparator/main/series_iterator_builder.go

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package main
+
+import (
+	"io"
+	"strings"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/encoding"
+	"github.com/m3db/m3/src/dbnode/encoding/m3tsz"
+	"github.com/m3db/m3/src/dbnode/namespace"
+	"github.com/m3db/m3/src/dbnode/x/xio"
+	"github.com/m3db/m3/src/x/ident"
+	xtime "github.com/m3db/m3/src/x/time"
+)
+
+const sep rune = '!'
+const tagSep rune = '.'
+
+type iteratorOptions struct {
+	blockSize     time.Duration
+	start         time.Time
+	encoderPool   encoding.EncoderPool
+	iteratorPools encoding.IteratorPools
+}
+
+var iterAlloc = func(r io.Reader, _ namespace.SchemaDescr) encoding.ReaderIterator {
+	return m3tsz.NewReaderIterator(r, m3tsz.DefaultIntOptimizationEnabled, encoding.NewOptions())
+}
+
+func buildBlockReader(
+	block seriesBlock,
+	start time.Time,
+	opts iteratorOptions,
+) ([]xio.BlockReader, error) {
+	encoder := opts.encoderPool.Get()
+	encoder.Reset(start, len(block), nil)
+	for _, dp := range block {
+		err := encoder.Encode(dp, xtime.Second, nil)
+		if err != nil {
+			encoder.Close()
+			return nil, err
+		}
+	}
+
+	segment := encoder.Discard()
+	encoder.Close()
```

Comment:
Discard actually also returns the encoder to the pool, so calling Close here would be a double close. (Yes Discard() is not super... ergonomic to use).

So just a single discard here needed.

Here's the relevant snippet:
```go
// Discard closes the encoder and transfers ownership of the data stream to
// the caller.
func (enc *encoder) Discard() ts.Segment {
	segment := enc.discard()

	// Close the encoder no longer needed
	enc.Close()

	return segment
}
```

Reply:
Ah nice, yeah

Reply:
Ah nice